File: Masjidly - Official Masjid Prayer Times/Domain/MosqueDefaults.swift
import Foundation

enum MosqueDefaults {
    static let defaultSlug = "muslim-welfare-house"

    static func visibleMosques(_ all: [Mosque]) -> [Mosque] {
        all.filter { !$0.isHiddenResolved }.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    static func resolveSelectedMosque(mosques: [Mosque], selectedId: String?, selectedSlug: String?) -> Mosque? {
        let visible = visibleMosques(mosques)
        guard !visible.isEmpty else { return nil }
        if let id = selectedId, let m = visible.first(where: { $0.id == id }) {
            return m
        }
        if let slug = selectedSlug, let m = visible.first(where: { $0.slug == slug }) {
            return m
        }
        return visible.first(where: { $0.slug == defaultSlug }) ?? visible[0]
    }

    // MARK: - Country

    /// Stable key for grouping mosques by country.
    static func countryGroupingKey(for mosque: Mosque) -> String {
        let code = mosque.countryCode?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return code.isEmpty ? "unknown" : code.uppercased()
    }

    /// Distinct countries for pickers (sorted by display label).
    static func countryOptions(from mosques: [Mosque]) -> [(key: String, label: String)] {
        let visible = visibleMosques(mosques)
        let grouped = Dictionary(grouping: visible, by: { countryGroupingKey(for: $0) })
        let keys = grouped.keys.sorted { lhs, rhs in
            countryLabel(for: lhs, grouped: grouped)
                .localizedCaseInsensitiveCompare(countryLabel(for: rhs, grouped: grouped)) == .orderedAscending
        }
        return keys.map { key in (key, countryLabel(for: key, grouped: grouped)) }
    }

    /// Filter mosques to those belonging to the given country grouping key.
    static func mosques(inCountryGroupingKey key: String, mosques: [Mosque]) -> [Mosque] {
        visibleMosques(mosques).filter { countryGroupingKey(for: $0) == key }
    }

    private static func countryLabel(for key: String, grouped: [String: [Mosque]]) -> String {
        grouped[key]?.first?.countryName
            ?? grouped[key]?.first?.countryCode
            ?? key
    }

    // MARK: - City

    /// Distinct cities for settings / onboarding pickers (sorted by display label).
    /// When `countryKey` is non-nil, only mosques within that country are considered.
    static func cityOptions(from mosques: [Mosque], countryKey: String? = nil) -> [(key: String, label: String)] {
        let filtered: [Mosque]
        if let countryKey, !countryKey.isEmpty {
            filtered = Self.mosques(inCountryGroupingKey: countryKey, mosques: mosques)
        } else {
            filtered = visibleMosques(mosques)
        }
        guard !filtered.isEmpty else { return [] }
        let grouped = Dictionary(grouping: filtered, by: { $0.cityGroupingKey })
        let keys = grouped.keys.sorted { lhs, rhs in
            cityLabel(for: lhs, grouped: grouped)
                .localizedCaseInsensitiveCompare(cityLabel(for: rhs, grouped: grouped)) == .orderedAscending
        }
        return keys.map { key in (key, cityLabel(for: key, grouped: grouped)) }
    }

    static func mosques(inCityGroupingKey key: String, mosques: [Mosque]) -> [Mosque] {
        visibleMosques(mosques).filter { $0.cityGroupingKey == key }
    }

    private static func cityLabel(for key: String, grouped: [String: [Mosque]]) -> String {
        grouped[key]?.first?.cityName
            ?? grouped[key]?.first?.cityDisplayName
            ?? key
    }
}

File: apps/expo/lib/prayer/mosqueDefaults.ts
import type { Mosque } from "@/types/prayer";

export const DEFAULT_MOSQUE_SLUG = "muslim-welfare-house";

/** Stable key for grouping / settings (matches native `Mosque.cityGroupingKey`). */
export function cityGroupingKey(m: Pick<Mosque, "citySlug" | "cityName">): string {
  if (m.citySlug && m.citySlug.length > 0) {
    return `slug:${m.citySlug}`;
  }
  const label = (m.cityName ?? "Sheffield").toLowerCase();
  return `name:${label}`;
}

// ── Country ──

/** Stable key for grouping mosques by country (matches native `MosqueDefaults.countryGroupingKey`). */
export function countryGroupingKey(m: Pick<Mosque, "countryCode">): string {
  const code = (m.countryCode ?? "").trim();
  return code.length > 0 ? code.toUpperCase() : "unknown";
}

export function countryOptions(mosques: Mosque[]): { key: string; label: string }[] {
  const visible = visibleMosques(mosques);
  const map = new Map<string, Mosque[]>();
  for (const m of visible) {
    const k = countryGroupingKey(m);
    if (!map.has(k)) map.set(k, []);
    map.get(k)!.push(m);
  }
  const keys = [...map.keys()].sort((a, b) =>
    countryLabelForKey(map, a).localeCompare(countryLabelForKey(map, b), undefined, {
      sensitivity: "base",
    })
  );
  return keys.map((key) => ({ key, label: countryLabelForKey(map, key) }));
}

function countryLabelForKey(grouped: Map<string, Mosque[]>, key: string): string {
  const list = grouped.get(key);
  const first = list?.[0];
  return first?.countryName ?? first?.countryCode ?? key;
}

export function mosquesInCountry(key: string, mosques: Mosque[]): Mosque[] {
  return visibleMosques(mosques).filter((m) => countryGroupingKey(m) === key);
}

// ── City ──

export function cityOptions(
  mosques: Mosque[],
  countryKey?: string
): { key: string; label: string }[] {
  const filtered = countryKey && countryKey.length > 0
    ? mosquesInCountry(countryKey, mosques)
    : visibleMosques(mosques);
  if (filtered.length === 0) return [];
  const map = new Map<string, Mosque[]>();
  for (const m of filtered) {
    const k = cityGroupingKey(m);
    if (!map.has(k)) map.set(k, []);
    map.get(k)!.push(m);
  }
  const keys = [...map.keys()].sort((a, b) =>
    cityLabelForKey(map, a).localeCompare(cityLabelForKey(map, b), undefined, {
      sensitivity: "base",
    })
  );
  return keys.map((key) => ({ key, label: cityLabelForKey(map, key) }));
}

function cityLabelForKey(grouped: Map<string, Mosque[]>, key: string): string {
  const list = grouped.get(key);
  const first = list?.[0];
  return first?.cityName ?? first?.citySlug ?? key;
}

export function mosquesInCity(key: string, mosques: Mosque[]): Mosque[] {
  return visibleMosques(mosques).filter((m) => cityGroupingKey(m) === key);
}

export function visibleMosques(all: Mosque[]): Mosque[] {
  return all
    .filter((m) => !m.isHiddenResolved)
    .sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: "base" }));
}

export function resolveSelectedMosque(
  mosques: Mosque[],
  selectedId: string | null | undefined,
  selectedSlug: string | null | undefined
): Mosque | null {
  const visible = visibleMosques(mosques);
  if (visible.length === 0) return null;
  if (selectedId) {
    const m = visible.find((m) => m.id === selectedId);
    if (m) return m;
  }
  if (selectedSlug) {
    const m = visible.find((m) => m.slug === selectedSlug);
    if (m) return m;
  }
  return visible.find((m) => m.slug === DEFAULT_MOSQUE_SLUG) ?? visible[0] ?? null;
}
